### ISSUE TITLE
Cryptic weaver error if database class extends non-database class

### ISSUE THREAD

**Issue by mtornwall** [Starcounter/Home#87]
Like the title says. Might be confusing for new users.

Repro:
```c#
using Starcounter;

public class Base { }
[Database] public class Derived : Base { }

class Program { static void Main() { } }
```

Weaver error:

```
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066: System.Exception: ScErrUnhandledWeaverException (SCERR6066): An unhandled, unidentified (non-Starcounter) exception occurred in the weaver.
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066: Version: 2.4.0.1636.
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066: Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR6066. ---> PostSharp.Sdk.CodeModel.BindingException: Cannot find a method named '.ctor' with signature '(uint16, [Starcounter]Starcounter.Binding.TypeBinding, [Starcounter]Starcounter.Internal.Uninitialized) : void' in scope 'Base'  (module = 'Bug.exe').
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    at PostSharp.Sdk.CodeModel.Collections.MethodDefDeclarationCollection.GetMethod(String name, IMethodSignature signature, BindingOptions bindingOptions)
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    at Starcounter.Internal.Weaver.ScTransformTask.EnhanceConstructors(TypeDefDeclaration typeDef, DatabaseClass databaseClass, TypeSpecificationEmit typeSpecification) in C:\TeamCity\TeamCity10\buildAgent\work\sc-4705\Level1\src\Starcounter.Weaver\Starcounter.Internal.Weaver\ScTransformTask.cs:line 858
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    at Starcounter.Internal.Weaver.ScTransformTask.Execute() in C:\TeamCity\TeamCity10\buildAgent\work\sc-4705\Level1\src\Starcounter.Weaver\Starcounter.Internal.Weaver\ScTransformTask.cs:line 307
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    at PostSharp.Sdk.Extensibility.Project.ExecutePhase(String phase)
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    at PostSharp.Hosting.PostSharpObject.ExecuteProjects()
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    at PostSharp.Hosting.PostSharpObject.ProcessAssemblies(ModuleLoadStrategy[] modules)
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    at Starcounter.Weaver.CodeWeaver.Execute() in C:\TeamCity\TeamCity10\buildAgent\work\sc-4705\Level1\src\Starcounter.Weaver\CodeWeaver.cs:line 257
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    at Starcounter.Weaver.CodeWeaver.ExecuteCurrent(CodeWeaver weaver) in C:\TeamCity\TeamCity10\buildAgent\work\sc-4705\Level1\src\Starcounter.Weaver\CodeWeaver.cs:line 82
1>C:\Program Files\Starcounter\Starcounter.MsBuild.targets(93,5): error SCERR6066:    --- End of inner exception stack trace ---
```

**Comment by per-samuelsson** [comment]
Wow, that's lovely. Not. 😃 

Thanks for reporting. I wonder if the experience is better in 2.3, but I doubt it.

Anyway, there are a few of these now, for cases that are not really covered in a UX-context, and like I raised in some meeting, we should probably take a look at error reporting between components in general. 

Also, the new weaver is under development, so another idea for these kind of weaver errors is we kick-start that one doing only front-level analysis, and take care of the reporting there, not even invoking the IL-based weaver if there are errors.

I'm not sure myself what approach to take. I can bring it up in some roadmap-discussion and we can figure it out together.

**Comment by per-samuelsson** [comment]
Tested this in 2.3 and found, to my surprise, that app start very well indeed. Meaning we support (by accident?) database classes whose base class is not. Need to think that through a bit.

**Comment by per-samuelsson** [comment]
Something clearly (not so) funny going on here. In 2.4, I can *neither* reproduce ***unless*** I make it a application project. Explicitly invoking the weaver happily weave the base class as a database class too.

Upgrading this to something we must fix.

**Comment by per-samuelsson** [comment]
> In 2.4, I can *neither* reproduce ***unless*** I make it a application project. Explicitly invoking the weaver happily weave the base class as a database class too.

Sloppy thinking. I remembered now that by default, assemblies compiled on the fly using `star <name>.cs` will be tagged `[Database]` on assembly level, meaning `Base` will pass as a database class, and weaving hence succeed.

Which is a relief: now we "only" need to be concerned with original UX.

**Comment by per-samuelsson** [comment]
Established this one applies to 2.3 too, so aiming to fix it in both while I'm at it.

**Comment by per-samuelsson** [comment]
Getting increasingly bananas that my fix works locally but fails with a really weird error on Simplified on TC servers, saying Content from Ring1 has a base class that ain't a database type (Something). Same tests work 100% locally, with same version of Simplified (I think at least). Anyway, taking longer than expect therefor.

**Comment by per-samuelsson** [comment]
> ... saying Content from Ring1 has a base class that ain't a database type (Something)

Wow, talk about being fooled: I misread, and there is not only a `Ring1.Content` (extending `Something`) in Simplified, there is also a `Ring1.Content` (extending `Entity`) in **SocietyObjects**. And build server pull SocietyObjects in in the background, and hence failure (my solution didn't consider `Entity` to be a valid base class - I thought it was deprecated/not in use since long).

**Comment by per-samuelsson** [comment]
Done. Merged into 2.3 in https://github.com/Starcounter/level1/commit/0699951a071ae32bda62aaee618ccb6f3f1e1a6d and then pushed. Then merged to **develop**.

**Comment by per-samuelsson** [comment]
Sample of new UX:

![image](https://user-images.githubusercontent.com/2902045/33667076-1c96356c-da9c-11e7-923f-60838a2ce3df.png)
